x86: fix QEMU accessing ACPI tables

dcpleung opened the pull request:
On QEMU with x86 and x86-64, there is a 1MB area immediately following SRAM where ACPI tables are stored. This area is not mapped by default and results in ACPI non-functional on QEMU. This PR maps this area by default so ACPI can be used on QEMU.

dcpleung committed:
boards: qemu: x86: run QEMU with correct memory size

QEMU x86 and x86_64 had total of 9MB which was for 4MB RAM,
4MB flash and 1MB of misc data (e.g. ACPI tables), where
flash area is only needed when CONFIG_XIP=y. However,
the gen_mmu.py only maps the 4MB RAM (and 4MB flash if XIP=y)
but not the last 1MB. This results in exceptions when
ACPI functions try to read those tables. This make
the memory size parameter to QEMU reflects what actually is
configured for the board with SRAM and flash.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: x86/info: also run qemu_x86 and qemu_x86_64

With the gen_mmu changes, tests/arch/x86/info can now be build
and run with qemu_x86 and qemu_x86_64. So add them to the list.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

tyrantlizard77 reviewed (commented):
I'm not a huge fan of modifying gen_mmu.py to add this case.
How early in the boot do we start probing this stuff? Is there any way we could add a run-time mapping as a PRE_KERNEL_1 init service in board-level code?

andrewboie commented:
also, what happens if we set the number of megabytes of memory QEMU has on the command line...to the same amount of RAM Zephyr thinks we have?

gen_mmu.py really is supposed to map all RAM, and we're hacking it because we're actually not.

dcpleung commented:
That actually was my original plan to map it in board init. However, AFAICT, `z_mem_map()` does not do identify mapping, so some plumbings need to be done in the ACPI layer. As those ACPI tables point to another set of tables where each would further point to some other tables, there will be lots of `z_mem_map()` done there. Not sure how well we can accommodate that many calls.

andrewboie commented:
> That actually was my original plan to map it in board init. However, AFAICT, z_mem_map() does not do identify mapping, so some plumbings need to be done in the ACPI layer. As those ACPI tables point to another set of tables where each would further point to some other tables, there will be lots of z_mem_map() done there. Not sure how well we can accommodate that many calls.

yeah I think the real issue is that we're not already mapping this RAM to begin with because QEMU and the Zephyr build have different values for how much RAM there actually is, we need to get them to be the same, and work on ways to not have the kernel and the firmware stomp on each other's memory

dcpleung committed:
x86: reserve memory for ACPI tables in linker script for QEMU

QEMU uses the last 128K of memory to store ACPI tables. So
shrink the SRAM/flash size to avoid placing anything in
this region.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
This now maps the whole SRAM/flash area while avoiding the last 128KB of memory.

tyrantlizard77 reviewed (commented):
Still don't think this scales to other x86 devices which might place the ACPI data at an unpredictable location. QEMU puts it at the end of RAM...but some other FW might put it somewhere else. How can we avoid such collisions in the general case?

Also I'd like to understand the life-cycle of the ACPI data. How late in the boot do we need it? For example, are we still querying it after the main thread starts? There may be a point in time where we no longer need this data and are free to overwrite it, but I am not sure.

Ultimately, it might be the case that we need to do E820 probing of the memory map and account for which physical pages are and are not usable, and have a concept of an anonymous page pool that gets memory-mapped on demand via _sbrk() calls instead of the RAM-sized physical region we have at the moment..

Not saying this is a -1 yet but if we go with this solution I'd still want a bug open to solve this in a robust way for all PC-like devices..

dcpleung commented:
This is strictly for QEMU so we can test ACPI code via CI.

If we go with this, I will file a bug on making `_sbrk()` aware of memory holes.

dcpleung committed:
tests: x86/info: enable testing for userspace

This adds some test cases to test with userspace enabled
to make sure will not be broken.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86: implement custom sbrk() for heap

The memory space in x86 is not necessarily linear. Most of
the time, there are various small regions being reserved,
for example, ACPI tables, BIOS, etc. So the default sbrk()
may result in memory allocated out of these regions, where
access is desired. So implement a custom sbrk() to restrict
where the end of heap is, as the heap area needs to be
continuous and can't go around these regions with current
code.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Updated with a custom `sbrk()`. Currently, it can't really work around holes as we need a linear, continuous memory space for heap. Once we have virtual memory, we can map all free memory into one continuous block.
